parser/src/grammar.lalrpop

```diff
@@ -0,0 +1,409 @@
+        let id = TypedIdent::new(id);
+        let args = args.into_iter()
+                       .map(TypedIdent::new)
+                       .collect();
+
+        Pattern::Constructor(id, args)
+    },
+};
+
+// Expressions
+
+pub Literal: Literal = {
+    "StringLiteral" => Literal::String(<>),
+    "CharLiteral" => Literal::Char(<>),
+    "IntLiteral" => Literal::Int(<>),
+    "ByteLiteral" => Literal::Byte(<>),
+    "FloatLiteral" => Literal::Float(<>),
+};
+
+Alternative: Alternative<Id> =
+    "|" <pat: S<Pattern>> "->" <expr: S<AtomicExpr>> => {
+        Alternative {
+            pattern: pat,
+            expr: expr,
+        }
+    };
+
+FieldExpr: FieldExpr<Id> = {
+    "type" <id: Ident> "=" <body : Type> =>
+        FieldExpr::Type(id, Some(body)),
+
+    "type" <id: Ident> =>
+        FieldExpr::Type(id, None),
+
+    <id: Ident> "=" <body: S<Expr>> =>
+        FieldExpr::Value(id, Some(body)),
+
+    <id: Ident> =>
+        FieldExpr::Value(id, None),
+};
+
+ValueBinding: ValueBinding<Id> = {
+    <comment: "DocComment"?> <name: S<AtomicPattern>> <typ: (":" <Type>)?> "=" <body: S<Expr>> =>
+        ValueBinding {
+            comment: comment,
+            name: name,
+            typ: typ.unwrap_or_else(Type::hole),
+            args: vec![],
+            expr: body,
+        },
+
+    <comment: "DocComment"?> <name: S<Ident>> <args: Ident+> <typ: (":" <Type>)?> "=" <body: S<Expr>> =>
+        ValueBinding {
+            comment: comment,
+            name: name.map(TypedIdent::new).map(Pattern::Ident),
+            typ: typ.unwrap_or_else(Type::hole),
+            args: args.into_iter().map(TypedIdent::new).collect(),
+            expr: body,
+        },
+};
+
+AtomicExpr: Expr<Id> = {
+    <id: Ident> =>
+        Expr::Ident(TypedIdent::new(id)),
+
+    <lit: Literal> =>
+        Expr::Literal(lit),
+
+    // "(" "." <id: Ident> ")" =>
+    //     Expr::Getter(id),
+
+    <expr: S<AtomicExpr>> "." <id: Ident> =>
+        Expr::Projection(Box::new(expr), id, Type::hole()),
+
+    "(" <elems: Comma<Expr>> ")" =>
+        match elems.len() {
+            0 => Expr::Tuple(vec![]),
+            1 => elems.into_iter().next().unwrap(),
+            _ => unimplemented!(),
+        },
+
+    "[" <elems: Comma<S<Expr>>> "]" => Expr::Array(Array {
+            typ: Type::hole(),
+            exprs: elems,
+        }),
+
+    "{" <fields: Comma<FieldExpr>> "}" => {
+        let mut types = Vec::new();
+        let mut values = Vec::new();
+
+        for field in fields {
+            match field {
+                FieldExpr::Type(id, typ) => types.push((id, typ)),
+                FieldExpr::Value(id, expr) => values.push((id, expr)),
+            }
+        }
+
+        Expr::Record {
+            typ: Type::hole(),
+            types: types,
+            exprs: values,
+        }
+    },
+
+    "OpenBlock" <exprs: (<S<Expr>> "Semi")*> <last: S<Expr>> "CloseBlock" => {
+        if exprs.is_empty() {
+            last.value
```

Comment:
I think this may give the wrong spans in some cases as it will give a span between `OpenBlock` and `CloseBlock` instead of the actual expression (didn't think about that when I wrote it.